python/metatensor-learn/metatensor/learn/data/collate.py

```diff
@@ -0,0 +1,56 @@
+"""
+Module containing a collate function for use in a Dataloader.
+"""
+
+import metatensor
+
+
+def collate_fn(batch, join: bool = False):
```

Comment:
So this would be the TensorMap collate? and for other data types we would have a different one?

Reply:
Good point. I suppose if the underlying data were `Systems` objects, for instance, the above collate function would work with `join=False`? I think I might have missed this in the discussion, but what are the other ways that one would want to collate samples that aren't TensorMaps?

Reply:
I think the general idea is the following: in the input is a TensorMap, or a torch.Tensor, we can join it (or not). For the rest in the inputs, we just return a subset of the original set of data (i.e. a sub-list or sub-dict).

It may be implemented in a following way (inspired by torch_geometric Dataloader)

```py
class DataLoader(torch.utils.data.DataLoader):
    def __init__(
        self,
        dataset: Union[Dataset, Sequence[BaseData]],
        batch_size: int = 1,
        shuffle: bool = False,
        **kwargs,
    ):

        self.collator = Collater(dataset) # A class for determining the behavior of the collate_fn, may be actually replaced with a single collate_fn, covering all use cases

        if isinstance(dataset, OnDiskDataset):
            dataset = range(len(dataset)) # in this case we do not collate the objects

        super().__init__(
            dataset,
            batch_size,
            shuffle,
            collate_fn=self.collator.collate_fn,
            **kwargs,
        )
```

So if the data type is recognised (i.e. it is a TensorMap, or a torch.Tensor the Collater will join the objects either by `metatensor.join()` or `torch.stack() / torch.cat()`). Otherwise it's just a list or dict of objects.

Reply:
Another question I have comes from looking at this example https://pytorch.org/tutorials/beginner/data_loading_tutorial.html . In this example they work with a dict dataset, and return the batches in the same dict format. While we do not need to replicate this necessary, it seems to me more intuitive. So in your notebook example would be
```python
batch = next(iter(train_dataloader)) # a dict with keys "descriptor", "target", "auxillary"
in_train_batch, out_train_batch, aux_train_batch = batch.values()
```
So here I get the same format out as I put in the data in the Dataset.

Reply:
So I've implemented 2 collate functions now, see comment below.